frontend/src/scenes/session-recordings/player/rrweb/index.ts

```diff
@@ -167,6 +167,77 @@ const defaultStyleRules = `.ph-no-capture { background-image: ${PLACEHOLDER_SVG_
 const shopifyShorthandCSSFix =
     '@media (prefers-reduced-motion: no-preference) { .scroll-trigger:not(.scroll-trigger--offscreen).animate--slide-in { animation: var(--animation-slide-in) } }'
 
+export type LogType = 'log' | 'warning'
+export type LoggingTimers = Record<LogType, NodeJS.Timeout | null>
+export type BuiltLogging = {
+    logger: playerConfig['logger']
+    timers: LoggingTimers
+}
+
+export const makeNoOpLogger = (): BuiltLogging => {
+    return {
+        logger: {
+            log: () => {},
+            warn: () => {},
+        },
+        timers: { log: null, warning: null },
+    }
+}
+
+export const makeLogger = (onIncrement: (count: number) => void): BuiltLogging => {
+    const counters = {
+        log: 0,
+        warning: 0,
+    }
+
+    ;(window as any)[`__posthog_player_logs`] = (window as any)[`__posthog_player_logs`] || []
+    ;(window as any)[`__posthog_player_warnings`] = (window as any)[`__posthog_player_warnings`] || []
+
+    const logStores: Record<LogType, any[]> = {
+        log: (window as any)[`__posthog_player_logs`],
+        warning: (window as any)[`__posthog_player_warnings`],
+    }
+
+    const timers: LoggingTimers = {
+        log: null,
+        warning: null,
+    }
+
+    const logger = (type: LogType): ((message?: any, ...optionalParams: any[]) => void) => {
+        // NOTE: RRWeb can log _alot_ of warnings,
+        // so we debounce the count otherwise we just end up making the performance worse
+        // We also don't log the messages directly.
+        // Sometimes the sheer size of messages and warnings can cause the browser to crash deserializing it all
+
+        return (...args: any[]): void => {
+            logStores[type].push(args)
+            counters[type] += 1
+
+            if (!timers[type]) {
+                timers[type] = setTimeout(() => {
+                    timers[type] = null
+                    if (type === 'warning') {
+                        onIncrement(logStores[type].length)
+                    }
+
+                    console.warn(
+                        `[PostHog Replayer] ${counters[type]} ${type}s (window.__posthog_player_${type}s to safely log them)`
```

Comment:
best to always comment on a file so we can thread :)

> Can we remove those by default as well?

Do you mean this line? It logged 6 times instead of hundreds so the future traveller can discover this log hiding

I don't think we should remove it, or folk won't know how to find logs if they want to look for them

Reply:
@veryayskiy